(WIP) v6

davidkpiano opened the pull request.

davidkpiano committed:
Support transition functions WIP

davidkpiano committed:
Enqueue actions WIP

davidkpiano committed:
Guards

davidkpiano committed:
Move enqueue obj

davidkpiano committed:
More tests

davidkpiano committed:
Add children to action arg, More tests

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Add evaluateCandidate, next set of tests

davidkpiano committed:
Add guards

davidkpiano committed:
Add more v6 tests

davidkpiano committed:
Move id counter to be per-system, add children to transition function

davidkpiano committed:
Reenter

davidkpiano committed:
WIP

davidkpiano committed:
Add some missing impls

davidkpiano committed:
Add test

davidkpiano committed:
Add invoke v6 tests

davidkpiano committed:
Add invoke v6 tests

davidkpiano committed:
Update snapshot

davidkpiano committed:
More tests

davidkpiano committed:
More tests

davidkpiano committed:
Rehydration tests WIP

davidkpiano committed:
Support for emitted events as actions and enq.action(fn, ...args)

davidkpiano committed:
Make sure to return new snapshot for transient transitions that update context

davidkpiano committed:
Add more test files

davidkpiano committed:
Add state.v6.test.ts

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Refactor state machine transition functions to use concise syntax

davidkpiano committed:
Refactor cont'd

davidkpiano committed:
Fix test

davidkpiano committed:
WIP

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
WIP

davidkpiano committed:
Actions WIP

davidkpiano committed:
Use vitest

davidkpiano committed:
Update assert tests

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
WIP

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Convert after tests

davidkpiano committed:
Remove assign.v6.tests.ts

davidkpiano committed:
Migrate tests

davidkpiano committed:
More tests

davidkpiano committed:
Guards

davidkpiano committed:
WIP

davidkpiano committed:
Fix test (pardon the dust)

davidkpiano committed:
Interpretertest

davidkpiano committed:
Invoke

davidkpiano committed:
Logger

davidkpiano committed:
Refactor log

davidkpiano committed:
sendTo WIP

davidkpiano committed:
Meta

davidkpiano committed:
Parallel & order

davidkpiano committed:
Spawn, spawnChild, state

davidkpiano committed:
stateIn, tags

davidkpiano committed:
Remove assign WIP

davidkpiano committed:
Replace createMachine WIP

davidkpiano committed:
WIP

davidkpiano committed:
schema.event-> schema.events

davidkpiano committed:
enq.action(…) -> enq(…)

davidkpiano committed:
Removing enqueueActions WIP

davidkpiano committed:
Cleanup & wip

davidkpiano committed:
Introduce src: ({ actors }) => …

davidkpiano committed:
Cleanup; remove xstate-immer package

davidkpiano committed:
Remove forwardTo

davidkpiano committed:
Remove sendParent

davidkpiano committed:
WIP

davidkpiano committed:
WIP: make actions tests pass

davidkpiano committed:
WIP actions tests

davidkpiano committed:
WIP

davidkpiano committed:
WIP

davidkpiano committed:
WIP

davidkpiano committed:
WIP remove createMachine

davidkpiano committed:
WIP

davidkpiano committed:
WIP fix tests

davidkpiano committed:
WIP fix tests

davidkpiano committed:
WIP

davidkpiano committed:
WIP: built-in actions

davidkpiano committed:
Fix delayed transitions

davidkpiano committed:
WIP state schema

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Some test fixes

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Skip rehydration, fix transient algo (WIP)

davidkpiano committed:
Simplify inspection

davidkpiano committed:
Update tests

davidkpiano committed:
Fix initial transitions

davidkpiano committed:
Cleanup

davidkpiano committed:
Get all tests passing (WIP)

davidkpiano committed:
Remove assign and enqueueActions

davidkpiano committed:
Fix typeHelpers.test.ts

davidkpiano committed:
Remove setup

davidkpiano committed:
Remove raise WIP

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Bulldozing action code

davidkpiano committed:
Remove raise!

davidkpiano committed:
Delete send action

davidkpiano committed:
Delete more actions (WIP cancel)

davidkpiano committed:
Remove guards

davidkpiano committed:
Remove all actions

davidkpiano committed:
Action logic cleanup

davidkpiano committed:
WIP cleanup

davidkpiano committed:
More cleanup

davidkpiano committed:
More cleanup

davidkpiano committed:
Type cleanup

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Fix test + package.json

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Undo TImplementation consolidation

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Cleanup actions.test.ts

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
createMachineFromConfig

davidkpiano committed:
Test cleanup

davidkpiano committed:
Add test from #5101

davidkpiano committed:
Remove __xstatenode

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Remove machine definition, types cleanup

davidkpiano committed:
Refactor StateMachine and StateNode types, removing unused parameters and cleaning up action handling. Update transition functions to use new ExecutableActionObject type. Clean up built-in actions and related type definitions

davidkpiano committed:
Enhance createMachine and StateNode functionality by introducing createStateConfig for improved state configuration. Clean up type definitions and remove unused parameters in createMachine and StateNode. Update getStateNodes to accept a more specific stateNode structure.

davidkpiano committed:
Add schemas for context and events in various tests using Zod for validation. Update existing tests to include new schema definitions for improved type safety and structure.

davidkpiano committed:
WIP React tests

davidkpiano committed:
Improve types

davidkpiano committed:
Tweak createMachineFromConfig

davidkpiano committed:
Remove createMachine

davidkpiano committed:
next_createMachine -> createMachine

davidkpiano committed:
Merge branch 'main' into v6

devinrhode2 commented:
I like tanstack routers type safe url support, but I hate unit tests more.

Whats the best we can integrate with tanstack router (or react router) with XState?

davidkpiano committed:
Change events and emitted schemas to schema maps

davidkpiano committed:
Refactor stateUtils WIP

davidkpiano committed:
TypeStates WIP

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Fix assert tests

davidkpiano committed:
Add trigger

davidkpiano committed:
Setup code coverage

davidkpiano committed:
Some cleanup

davidkpiano committed:
Cleanup

davidkpiano committed:
Simplify initial transitions (not dynamic)

davidkpiano committed:
WIP

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Fix tests

davidkpiano committed:
fn -> to

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Fix SCXML WIP

davidkpiano committed:
WIP: fix send/raise events

davidkpiano committed:
WIP SCXML

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Cleanup WIP

davidkpiano committed:
Make actors thenable

davidkpiano committed:
Revert "Make actors thenable"

This reverts commit addefe987173860778a86e14ea72c2c29449e857.

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Add enq.listen and enq.subscribeTo

davidkpiano committed:
Add logic.createActor + tests

davidkpiano committed:
State params

davidkpiano committed:
Strongly typed state params

davidkpiano committed:
Update gitignore

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Remove unused types WIP

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Fix routable state types

davidkpiano committed:
Fix SCXML tests

davidkpiano committed:
Fix types WIP

davidkpiano committed:
Add TODO

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Add internalEvents

davidkpiano committed:
Refactor createMachine and setup interfaces to use 'const' for TEventSchemaMap type parameter

davidkpiano committed:
Get framework tests to pass & remove xstate-inspect

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Add stateNode to delays

davidkpiano committed:
Parallel aggregation

davidkpiano committed:
Context type inference from config.context when no schema provided

davidkpiano committed:
Fix internalEvents

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Fix tests

davidkpiano committed:
State params -> input (rename)

davidkpiano committed:
inputSchema -> schemas.input

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Add triggers

davidkpiano committed:
Remove .createActor and make invoke.src resolve actor logic

Strongly type string invoke sources against actor keys when an actors
object is present, while still allowing plain strings when it is not.
This also removes the .createActor convenience in favor of
createActor(logic, options).

davidkpiano committed:
Add state and invoke timeouts

davidkpiano committed:
Add choice state routing

davidkpiano committed:
Cleanup

davidkpiano committed:
Support ISO8601 duration parsing

davidkpiano committed:
refactor(core): simplify delayed transition normalization

davidkpiano committed:
refactor(core): simplify StateNode transition mapping

davidkpiano committed:
refactor(core): share delay parsing helpers

davidkpiano committed:
refactor(core): simplify delayed event scheduling

davidkpiano committed:
refactor(core): streamline state node collection

davidkpiano committed:
refactor(core): streamline transition state checks

davidkpiano committed:
refactor(core): reduce transition conflict allocations

davidkpiano committed:
refactor(core): streamline child state iteration

davidkpiano committed:
refactor(core): add hasDescendantState utility for improved state checks

davidkpiano committed:
refactor(core): reduce state utils allocations

davidkpiano committed:
refactor(core): streamline state utils transitions

davidkpiano committed:
refactor(core): reduce state utils helper overhead

davidkpiano committed:
refactor(core): simplify microstep logic and remove unused functions

davidkpiano committed:
refactor(core): localize state node helpers

davidkpiano committed:
refactor(core): simplify state node event getters

davidkpiano committed:
refactor(core): streamline state node initialization

davidkpiano committed:
refactor(core): reduce StateNode memoization

davidkpiano committed:
refactor(core): streamline StateNode event handling

davidkpiano committed:
refactor(core): simplify machine snapshot restoration

davidkpiano committed:
refactor(core): streamline delayed transition formatting

davidkpiano committed:
refactor(core): dedupe transition config formatting

davidkpiano committed:
refactor(core): simplify delay helpers

davidkpiano committed:
refactor(core): reduce state utils indirection

davidkpiano committed:
refactor(core): dedupe transition effect checks

davidkpiano committed:
refactor(core): simplify transition node selection

davidkpiano committed:
refactor(core): simplify transition domain lookup

davidkpiano committed:
refactor(core): remove dead initial transition helper

davidkpiano committed:
refactor(core): reduce state utils helper indirection

davidkpiano committed:
refactor(core): streamline actor notifications

davidkpiano committed:
refactor(core): simplify actor completion state

davidkpiano committed:
refactor(core): reduce actor error handling indirection

davidkpiano committed:
refactor(core): reduce snapshot can allocations

davidkpiano committed:
refactor(core): streamline snapshot metadata

davidkpiano committed:
refactor(core): streamline snapshot checks

davidkpiano committed:
refactor(core): reuse snapshot state values

davidkpiano committed:
refactor(core): short-circuit targeted can checks

davidkpiano committed:
refactor(core): reduce microstep collection copies

davidkpiano committed:
Rename fromPromise to createLogic

davidkpiano committed:
Add createLogic actor helpers

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
Fix migration.md

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
SCXML tests

davidkpiano committed:
Fix dual Vue instance in tests by pinning vue@3.5.17

@vue/test-utils was resolving to vue@3.5.33 while test code used
vue@3.5.17, causing two separate Vue reactivity systems and preventing
DOM updates in Vue store tests.

davidkpiano committed:
Merge branch 'main' into v6

davidkpiano committed:
XState atoms

davidkpiano committed:
Fix v6 release readiness checks

davidkpiano committed:
refactor(v6): engine/actor/atom fixes + cross-package dedupe

Audit-driven improvements for v6. All non-breaking except the DoneActorEvent
type tightening (acceptable for a major).

Engine:
- Fix invoked-flag clone bug in enterStates: the flag was set once and never
  reset, so cloneMachineSnapshot re-ran for every node entered after the first
  invoke even when children were unchanged. Now scoped per node.
- getAllStateNodes: early-terminate the ancestor walk when a node is already in
  the set, so shared ancestor chains aren't re-walked once per sibling
  (O(n*depth) -> ~O(n+depth)). Safe because Set iteration visits nodes added
  mid-loop, so every chain is still fully covered.

Actors:
- Clear _deferred in the init-failure catch so functions deferred during a
  throwing getInitialSnapshot/restoreSnapshot can't run against an error actor.

Atoms (core):
- createAsyncAtom: default to a value-based compare (status + Object.is on
  data/error) so pending->pending and identical results no longer emit
  spurious notifications.
- Atom.subscribe: route observer throws to reportUnhandledError (try/catch)
  instead of letting them break reactive propagation, matching Actor.subscribe.

Types:
- InvokeConfig.onDone: DoneActorEvent<any> -> DoneActorEvent<unknown>, aligning
  with the type's own defaults and stopping unchecked event.output flow.

Cleanup:
- Export isAtom from @xstate/store and remove the 4 byte-identical copies in
  the react/preact/vue/solid bindings (-32 LOC, single source of truth).
- Remove dead optionsRef (+ orphaned useRef import) from useIdleActorRef.

scxml.ts: documented as @internal (test-only, not public API) per maintainer.

davidkpiano committed:
fix(v6): reject invalid action implementation shapes

The `Implementations['actions']` constraint was `(...args: any[]) => void`,
which let any function be registered as a named action with zero compile-time
feedback — e.g. `foo: (n: number) => 'hello'` type-checked despite returning a
value an action can never return, then silently misbehaved at runtime.

Tighten the action return type to `void | { context?, children? }` so the
definition site rejects functions that don't return a valid action result.
Add a focused negative test at the createMachine site.

davidkpiano committed:
fix(v6): close raise/sendTo gaps in getTransitionResult enqueue

The enqueue object used to evaluate `to:` transition functions diverged from
the one used in microstep:

- raise() skipped the string-event validation, so enq.raise('SOME_EVENT') in a
  transition function silently pushed a bare string as an internal event
  instead of throwing the helpful "use raise({ type: ... })" error that entry/
  exit actions already produce.
- sendTo() pushed an @xstate.sendTo action even when the target actorRef was
  undefined, unlike microstep which no-ops. Now guarded to match.

Add transition() tests covering both.

Note: the delayed-raise path still differs intentionally (getTransitionResult
schedules inline vs microstep's @xstate.raise builtin); unifying that would be
a behavior change and is left out of this fix.

davidkpiano committed:
fix(v6): align store useSelector compare signature with actor binding

The react/preact/solid store bindings typed `compare` (and `defaultCompare`)
as `(a: T | undefined, b: T)`, but every call site guards `previous !==
undefined` before invoking it, so the `undefined` was never actually passed —
a dishonest signature that produced spurious TS errors when sharing a compare
function with @xstate/react's actor `useSelector`, which uses `(a: T, b: T)`.

Standardize on `(a: T, b: T)`, matching the actor binding and the already-
correct vue/svelte/angular store bindings. Existing `(a: T | undefined, b: T)`
user compares remain assignable.

davidkpiano committed:
refactor(v6): bundle size, invoke input typing, choice states, serialization

- Shrink the minimal createMachine+createActor bundle 16.8 → 14.8 kB min+gz
  (−12%): decouple atom/alien reactivity from createActor via a lazily
  installed interop hook, dedupe listener/subscription logic through a shared
  attached-actor factory, slim machineConfigToJSON, and gate long diagnostic
  messages to development builds. Add scripts/bundle-size.mjs with CI
  thresholds and a --why per-module attribution flag.
- Typecheck invoke `input` (static values and mappers) against the registered
  actor logic's input type by distributing the invoke config over the
  `actors` map; `src: ({ actors }) => actors.x` resolvers are correlated too.
- Choice states: rename `choices` → `choice`, function-only (the
  array-of-guarded-targets form is removed). A choice state declares one
  function that must resolve to a target.
- Machines as data: spawned actors register on snapshot.children and persist
  like invoked actors; invoked actors keep their logical `src` in persisted
  snapshots; rehydrated children restart on start(); machine.toJSON() /
  .definition emit JSON-safe definitions with $unserializable markers that
  round-trip through createMachineFromConfig.
- Add the first-ten-minutes DX benchmark suite and a type-error-quality
  harness (scripts/type-error-quality.mjs + fixtures).

davidkpiano committed:
WIP

davidkpiano committed:
Update gitignore

davidkpiano committed:
chore(v6): clear eslint errors for release (B4)

Clears all pre-existing lint errors so the CI lint step passes:
- core/src: remove redundant type assertions, @ts-ignore→@ts-expect-error,
  prefix unused generics/args, guard base-to-string in error fallbacks,
  brace case-block lexical decls, drop now-unused ActorSrcKey
- xstate-react: drop redundant createActor(...) as Actor<TLogic> casts
- xstate-svelte test fixture: async-without-await
- xstate-store-react/src/test.tsx (stray SWAPI demo): suppress misused-promises

No runtime/behavior changes.

davidkpiano committed:
feat(v6): type-only schemas + xstate migrate codemod CLI

- Add types<T>() for type-only schemas (inference, no runtime validation) —
  the v6 replacement for v5's types: {} as {...}. Exported with isTypeSchema.
- Add @xstate/codemod + a zero-dep `xstate migrate` launcher in core that
  delegates via npx (so `npm i xstate` stays dependency-free). First release
  automates Tier A renames (interpret/Interpreter, from*->create*Logic),
  flags manual sites; fromPromise left for manual (shape change).
- Exclude packages/codemod from preconstruct (it's a CLI, not a bundled lib).

davidkpiano committed:
Update gitignore

davidkpiano committed:
feat(v6): atoms as reactive subscription input + lossless inspection protocol

Atoms as machine input:
- enq.subscribeTo() now accepts an atom directly (createAtom /
  createReducerAtom / createAsyncAtom); the mapper receives the atom's value.
  Adds the `isAtom` brand + predicate (exported).
- Attached subscription/listener actors tear down on parent stop and via
  enq.stop(); fix special actions returning only `{ children }` (e.g.
  registerSpawnedChild/unregisterChild) wiping context to undefined.

Inspection:
- Lossless inspection protocol (createActor/inspection/system) with
  conformance tests and changeset.

Docs:
- migration.md updated for atoms/reactive input, actor.select/get, route
  states, the actor registry, snapshot versioning, and serialization.

davidkpiano committed:
Optimize algo size WIP

Andarist committed:
Enter prerelease mode (#5542)

davidkpiano committed:
Merge branch 'v6' of https://github.com/statelyai/xstate into v6

davidkpiano commented:
Superseded by #5543
